### PR TITLE
Graveler cache repository information

### LINKED ISSUE #4425: Cache lakeFS repository object
It is currently being fetched by lakeFS for every request (batched for 3ms, still a lot). It's not editable. We should cache it for short time (<5 seconds) to reduce the amount of calls to the KV store.

### PR DESCRIPTION
**Pull request by nopcoder**
Close #4425 

Cache repository information in Graveler ref manager kv implementation.

### CODE REVIEW

File: pkg/catalog/catalog.go [hunk 1]
```diff
@@ -107,9 +107,10 @@ const (
 )
 
 type Config struct {
-	Config        *config.Config
-	KVStore       *kv.StoreMessage
-	WalkerFactory WalkerFactory
+	Config             *config.Config
+	KVStore            *kv.StoreMessage
+	WalkerFactory      WalkerFactory
+	UseRepositoryCache bool
```
Comment (N-o-Z) [inline pkg/catalog/catalog.go:113]: Why is this a configurable value? Where do you set it to true (other than tests)

**Review by N-o-Z — COMMENTED** [review]
Cool addition,
though I didn't quite understand where we enable it in the server.
Also, not quite sure this should be configurable in catalog. If it was just for test enabling purposes, I think there are better options for that

**Comment by nopcoder** [comment]
> Cool addition, though I didn't quite understand where we enable it in the server. Also, not quite sure this should be configurable in catalog. If it was just for test enabling purposes, I think there are better options for that

Here https://github.com/treeverse/lakeFS/pull/4439/files#diff-dd51eee5d8ad1d57c5e785105ca5a0556ed5e2119cf0173e111f992e80a4dcb9R84 we set the default which is true.
At first I've enable the configuration of with/without cache at the ref manager level for testing.
But, when we test the API in the controller's tests we verify that when we delete a repository, we will get not found on the following get request.
In order to support this test and the fact that we setup the test by configure a local handler that setup the our catalog as part of the implementation - I've added a configuration option at the catalog level.
There https://github.com/treeverse/lakeFS/pull/4439/files#diff-13e18d34ca6c711c206f908ee31c4ca09844bd36684bb628b22fa994e2b940eeR180 we also set the default to true for the catalog level.

File: pkg/catalog/catalog.go (hunk 1 above)
Comment (nopcoder) [reply pkg/catalog/catalog.go:113]: Added comment explaining why we have a catalog level configuration that configure the ref manager's level.

Comment (N-o-Z) [reply pkg/catalog/catalog.go:113]: Why are we not just passing a cache object to the ref manager? That is our convention for managers for the most part

Comment (nopcoder) [reply pkg/catalog/catalog.go:113]: At this level for simplicity as we are going to add more caches, enabling it using boolean looks like a good first step.
Next we can configure each one using application level settings.

File: pkg/graveler/ref/manager_test.go [hunk 2]
```diff
@@ -12,14 +12,39 @@ import (
...
+	"github.com/treeverse/lakefs/pkg/batch"
 	"github.com/treeverse/lakefs/pkg/graveler"
+	"github.com/treeverse/lakefs/pkg/graveler/ref"
 	"github.com/treeverse/lakefs/pkg/ident"
 	"github.com/treeverse/lakefs/pkg/kv"
+	"github.com/treeverse/lakefs/pkg/kv/mock"
 	"github.com/treeverse/lakefs/pkg/testutil"
 )
 
+// TestManager_GetRepositoryCache test get repository information while using cache. Match the number of times we
```
Comment (N-o-Z) [inline pkg/graveler/ref/manager_test.go:27]: Maybe we can add to the test another call after 5 seconds and verify that `Get` was called again?

File: pkg/api/controller_test.go [hunk 3]
```diff
@@ -924,9 +925,12 @@ func TestController_DeleteRepositoryHandler(t *testing.T) {
 		resp, err := clt.DeleteRepositoryWithResponse(ctx, "my-new-repo")
 		verifyResponseOK(t, resp, err)
 
+		// wait for cache expiry
+		time.Sleep(ref.RepositoryCacheExpiry + ref.RepositoryCacheJitter + time.Second)
```
Comment (itaiad200) [inline pkg/api/controller_test.go:929]: That's a lot of time to sleep in a single test. Let's make it configurable for the tests?

**Review by N-o-Z — APPROVED** [review]
Looks great!
One suggestion regarding test scenario.
One comment regarding the formatting changes - I think it would be much better in the future to do these changes in a separate issue, as it makes it much harder to concentrate on the actual changes.
Thanks!

File: pkg/graveler/ref/manager.go [hunk 4]
```diff
@@ -58,46 +66,58 @@ func protoFromBranch(branchID graveler.BranchID, b *graveler.Branch) *graveler.B
...
+			return nil, err
+		}
+
+		switch repo.State {
+		case graveler.RepositoryState_ACTIVE:
+			return repo, nil
+		case graveler.RepositoryState_IN_DELETION:
+			return nil, graveler.ErrRepositoryInDeletion
+		default:
+			return nil, fmt.Errorf("invalid repository state (%d) rec: %w", repo.State, graveler.ErrInvalid)
```
Comment (itaiad200) [inline pkg/graveler/ref/manager.go:114]: Is a number for a state indicative in the returned error?

File: pkg/graveler/ref/manager_test.go [hunk 5]
```diff
@@ -161,6 +186,9 @@ func TestManager_DeleteRepository(t *testing.T) {
 			t.Fatalf("unexpected error: %v", err)
 		}
 
+		// wait for cache expiry
+		time.Sleep(ref.RepositoryCacheExpiry + ref.RepositoryCacheJitter + time.Second)
```
Comment (itaiad200) [inline pkg/graveler/ref/manager_test.go:190]: So long....

File: pkg/api/controller_test.go (hunk 3 above)
Comment (itaiad200) [reply pkg/api/controller_test.go:929]: Also if delete repo isn't using the cache then why wait?

**Review by itaiad200 — APPROVED** [review]
Please avoid sleeping in tests

File: pkg/graveler/ref/manager_test.go (hunk 5 above)
Comment (N-o-Z) [reply pkg/graveler/ref/manager_test.go:190]: Could be avoided if provided a unique repo name

File: pkg/api/controller_test.go (hunk 3 above)
Comment (N-o-Z) [reply pkg/api/controller_test.go:929]: I don't think it's that bad to add this sleep to only one test. Feels like adding a configuration just for the sake of a single test is a far worse option

Comment (nopcoder) [reply pkg/api/controller_test.go:929]: As long as there is no way to control the cache time or enable/disable - will need to wait while working with GetRepository.
Can make DeleteRepository skip the cache mechanism and write the test using two DeleteRepository calls to remove the sleep here.

File: pkg/graveler/ref/manager.go (hunk 4 above)
Comment (nopcoder) [reply pkg/graveler/ref/manager.go:114]: 1. Just made the format explict - it was a number
2. When we will get this error it means we have a bug and we should handle the new repository state here or add a new error UnknownState and drop this format.
3. 
Let me know what you think I will handle it in a different PR.

File: pkg/graveler/ref/manager_test.go (hunk 5 above)
Comment (nopcoder) [reply pkg/graveler/ref/manager_test.go:190]: Can't be avoided - this is the test - delete + verify + create again.

Comment (itaiad200) [reply pkg/graveler/ref/manager_test.go:198]: We can have the cache configuration injected and configure it to 0 for tests.

**Comment by nopcoder** [comment]
Configured cache parameters as part of ref manager to enable less sleep time while running tests.

**Pull request merged**